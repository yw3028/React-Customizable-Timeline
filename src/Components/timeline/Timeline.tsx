// eslint-disable-next-line no-use-before-define
import React from 'react';
import { useSpring, animated, config } from 'react-spring';
import { ThemeProvider } from 'styled-components';
import TimelineItem from '../timelineItem/TimelineItem';
import { STimeline, STimelineSection, STitle } from '../styles';
import { Item } from '../../Models/Item';
import { Fade } from 'react-awesome-reveal';

type Props = {
  animation?: boolean;
  data?: Item[];
  primaryDarkColor?: string;
  primaryLightColor?: string;
  primaryFont?: string;
  titleShape?: string;
  dotShape?: string;
  lineStyle?: string;
};

function Timeline({
  animation = true,
  data,
  primaryDarkColor = '#625261',
  primaryLightColor = '#F5F5DC',
  primaryFont = "'Chivo', sans-serif",
  titleShape = 'circle',
  dotShape = 'circle',
  lineStyle = 'dotted',
}: Props) {
  const theme = {
    primaryDarkColor,
    primaryLightColor,
    primaryFont,
    titleShape,
    dotShape,
    lineStyle,
  };

  const titleAnimation: Object = useSpring({
    from: {
      transform: `translate3d(0, -4px, 0) scaleY(1.02)`,
    },
    to: {
      transform: 'translate3d(0, 0, 0) scaleY(1.02)',
    },
    config: config.wobbly,
  });

  const lineAnimation: Object = useSpring({
    from: { transform: 'translate3d(-100%, 0, 0)' },
    to: { transform: 'translate3d(0, 0, 0)' },
    config: { mass: 2, tension: 120, friction: 24 },
    reset: animated,
  } as any);

  return (
    <ThemeProvider theme={theme}>
      <animated.div style={animation ? lineAnimation : undefined}>
        <STimeline>
          {data?.map((item, index) => {
            return (
              <STimelineSection key={index}>
                <animated.div style={animation ? titleAnimation : undefined}>
                  <STitle key={item.title}>
                    <div>{item.title}</div>
                  </STitle>
                </animated.div>
                {item.events.map((event, index) => {
                  return (
                    <Fade key={index}>
                      <TimelineItem
                        key={index}
                        title={event.title}
                        subtitle={event.subtitle}
                        content={event.content}
                        img={event.img}
                        label={event.label}
                        location={event.location}
                        animation={animation}
                      />
                    </Fade>
                  );
                })}
              </STimelineSection>
            );
          })}
        </STimeline>
      </animated.div>
    </ThemeProvider>
  );
}

export default Timeline;
