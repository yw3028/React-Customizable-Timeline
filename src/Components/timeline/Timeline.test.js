/* eslint-disable no-undef */
import { cleanup, render, screen } from '@testing-library/react';

import React from 'react';
import Timeline from './Timeline';
import data from '../mocks/index';
import IntersectionObserver from '../helper/testHelper';

global.IntersectionObserver = IntersectionObserver;

const renderComponent = data => render(<Timeline data={data} />);

afterEach(() => {
  cleanup();
});

describe('Timeline component', () => {
  it('render correctly', async () => {
    const component = renderComponent(data);
    await expect(component.baseElement).toMatchSnapshot();
  });

  it('Renders the title', async () => {
    renderComponent(data);
    await expect(screen.getByText('Avengers')).toBeTruthy();
  });

  // eslint-disable-next-line quotes
  it("Doesn't render children elements", async () => {
    render(
      <Timeline data={data}>
        <p>LisaPisa</p>
      </Timeline>,
    );

    await expect(screen.queryByText(/LisaPisa/i)).toBeFalsy();
  });
});
