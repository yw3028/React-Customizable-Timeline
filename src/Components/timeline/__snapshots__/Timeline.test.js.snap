// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Timeline component render correctly 1`] = `
<body>
  <div>
    <div
      style="transform: translate3d(-100%, 0, 0);"
    >
      <div
        class="sc-bwzfXH kYoUhp"
      >
        <div
          class="sc-htpNat fjcJqy"
        >
          <div
            style="transform: translate3d(0, -4px, 0) scaleY(1.02);"
          >
            <div
              class="sc-bdVaJa btHhJ"
            >
              <div>
                Avengers
              </div>
            </div>
          </div>
          <div
            class="css-q7lffx"
          >
            <div
              class="sc-bxivhb hkpChk"
            >
              <div
                style="transform: scale3d(1.25, 0.75, 1);"
              >
                <div
                  style="cursor: pointer;"
                >
                  <h4>
                    Avengers:StartGame
                  </h4>
                  <h5>
                    StartGame
                  </h5>
                  <img
                    alt="picture"
                    src="https://images.unsplash.com/photo-1600790078201-5490baf711d6?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=500&q=60"
                    style="max-width: 400px; width: 100%; min-width: 100px; padding-bottom: 10px;"
                  />
                </div>
                <div
                  style="font-size: small; text-align: right;"
                >
                  <span
                    aria-label="localtion"
                    role="img"
                  >
                    üìç
                  </span>
                   
                  Knowhere
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
